### ISSUE TITLE
Views UI: Fix "Replacement patterns" visibility for Global: Custom Text

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1928]
In the views UI when adding a global text field I see the following description text:

```
You may enter data from this view as per the "Replacement patterns" below.
```

But there is no "Replacement patterns" below.

----
PR: https://github.com/backdrop/backdrop/pull/3014 by @herbdool
PR: https://github.com/backdrop/backdrop/pull/3828 by @quicksketch

**Comment by klonos** [comment]
It's there for me:

![backdrop-views-add_global_text_field-patterns](https://cloud.githubusercontent.com/assets/2423362/15540054/02c58070-22c9-11e6-9b1e-90ed231aaf9e.gif)

**Comment by jenlampton** [comment]
huh. I wonder if it's my site. I'll spin up a demo and see if I can reproduce it.

**Comment by herbdool** [comment]
I have the same issue. It's on a site that's been upgraded from Drupal 6 so I wonder if that's the difference.

**Comment by olafgrabienski** [comment]
On my test site which isn't a former D6 site, the Replacement patterns didn't show up either.

In Firebug, I noticed a `display: none;` on the element:

![screen-backdrop-view-global-custom-text-replacement-patterns-display-none](https://cloud.githubusercontent.com/assets/14188307/21327910/95f9a9aa-c631-11e6-8750-54976f0fc3f9.png)

**Comment by olafgrabienski** [comment]
> But then I saved the view and went back to edit it and the Replacement Patterns accordion shows up.

Confirmed!

**Comment by bradbulger** [comment]
When I create a Global: Custom text field the Replacement Patterns fieldset is not there, unless I go check the "Output this field as a link" box in the Rewrite Results fieldset.

**Comment by jenlampton** [comment]
@bradbulger I don't think the replacement patterns are intended to work with a `Global: Custom text` field. Do they work if you use the tokens from the `Output this field as a link` section? If so, that's a handy feature, and we should make those patterns available all the time :)

**Comment by bradbulger** [comment]
Oh I thought that's what this meant. Yes, the text under the field says
"The text to display for this field. You may include HTML. You may enter
data from this view as per the "Replacement patterns" below." This is a
long standing workaround to "I need a more complicated display" in Views.
They work find, it's just about getting them to show up. I think that's a
long standing issue, too, I know I'm quite used to having to go check
something else when editing a field in Views in order to get the list of
field tokens to display.

**Comment by jenlampton** [comment]
Ah, well then perhaps it is related (I was thinking maybe it wasn't). If we can find where that `display: none` is coming from, and remove it, we can fix all the things!

**Comment by BWPanda** [comment]
The `display: none;` comes from the Form API `#states`. See: https://github.com/backdrop/backdrop/blob/e41153fbe25ccdb7a5b59894a965aca871cf7993/core/modules/views/handlers/views_handler_field.inc#L860

This definitely could use some work. Not only for the 'Global: Custom Text' field, but also because I don't think a fieldset's visibility should rely on checkboxes inside a different fieldset...

**Comment by jenlampton** [comment]
This might be leftover from removing the ctools states.

**Comment by stpaultim** [comment]
I was just looking for replacement patterns in a Global Text field in views. How can I help fix this?

**Comment by herbdool** [comment]
In the spirit of keeping the fix focused, I've added another condition to the states array that @BWPanda noted. If the Global Custom Text field is used and filled in then it will display the Replacement Patterns. I was also considering also having it check if the Global Custom Text field is empty, but I wasn't sure if it would then show up for all fields. I think I'll check.

This issue is a bug fix specific to the Global Custom Text field so I think @BWPanda should create follow-up feature/enhancement requests. Those requests aren't as simple as the bug fix.

**Comment by herbdool** [comment]
As suspected, testing a state where the custom text field is empty `':input[name="options[alter][text]"]' => array('empty' => TRUE)` results in it always showing for other fields, even if the custom text field isn't included in the view. 

I also tried `':input[name="options[alter][text]"]' => array('filled' => FALSE)` and `':input[name="options[alter][text]"]' => array('value' => '')`. Those also had the same result.

**Comment by klonos** [comment]
The code changes seems simple; thanks @herbdool 👍  ...can someone please add simple steps to reproduce this, so that I can test?

**Comment by herbdool** [comment]
@klonos create a new view using fields. Add a custom text field. When entering text the Replacement Patterns fieldset should appear.

You can also create a separate view of nodes with fields. In the title field click Rewrite output (IIRC) and the RP fieldset should appear only when it's clicked.

**Comment by olafgrabienski** [comment]
@herbdool Both ways work for me: *Rewrite the output of this field*, and also the *Global: Custom text* field.

Regarding the latter, the "Replacement Patterns" fieldset appears indeed only when text is entered in the textarea. I found it surprising because the behavior differs from "Rewrite output" where the patterns fieldset is visible all the time. Apart of that, the help in the Custom Text field mentions the replacement patterns being "below" in any case, that's a bit confusing.

**Comment by klonos** [comment]
> ... create a new view using fields. Add a custom text field. When entering text the Replacement Patterns fieldset should appear.

This works for me even w/o this fix (I mentioned this in a previous comment back in 2016). What am I missing?

> You can also create a separate view of nodes with fields. In the title field click Rewrite output (IIRC) and the RP fieldset should appear only when it's clicked.

...also works w/o this fix.

PS: I've tested by comparing behavior in the PR sandbox vs a tugboat demo site.

**Comment by herbdool** [comment]
@klonos it has to be a new, **unsaved** view for testing.

**Comment by stpaultim** [comment]
@klonos - You are correct. I have had this problem before and tried to recreate it in a Tugboat site and did NOT have the problem initially, but figured out how to recreate it. 

1) Open a new Tugboat site
2) Create a new view of Content, Type = All, Display format = unformatted list of fields
3) Add body field to views. (Default settings fine)
4) Add "Global: Custom Text" field
5) Check for "Replacement Patterns" and they will NOT be there. (BROKEN)

It's weird, because if you add the Global: Custom Text field before the body field, it works fine. It also works fine after you save and edit the view. 

But, if you try to add the Global: Custom Text field after adding another field first (and before saving the view) the Replacement Patterns fieldset is not available.

**Comment by herbdool** [comment]
@stpaultim I tested again, using your steps. It still works for me on the pr site. It appears when you start typing in the textures.

**Comment by stpaultim** [comment]
@herbdool - I tested again as well and you are correct. The `Replacement Patterns` field does show up once you start typing. I find it really odd that if follow these steps (in PR sandbox):

1) Create a new view of Content, Type = All, Display format = unformatted list of fields
2) Add `Global: Custom Text` field (`Replacement Patterns` field set is visible without any action).
3) Add body field to views. (Default settings fine)
4) Add "Global: Custom Text" field (`Replacement Patterns` field set is NOT visible).
5) Start typing in `Text` field and the `Replacement Patterns` field set appears. 

This is an improvement, but it's still inconsistent and a little flakey. Do you have any idea of why it's behaving differently under these circumstances and/or which is the expected behavior (show up immediately or only after typing)?

**Comment by olafgrabienski** [comment]
> it's still inconsistent and a little flakey. Do you have any idea of why it's behaving differently under these circumstances

I'm also interested in an answer to @stpaultim's question about a reason for the inconsistent behavior.

In case you didn't notice [my respective comment above](https://github.com/backdrop/backdrop-issues/issues/1928#issuecomment-565100832):

>> ... the "Replacement Patterns" fieldset appears indeed only when text is entered in the textarea. I found it surprising because the behavior differs from "Rewrite output" where the patterns fieldset is visible all the time. Apart of that, the help in the Custom Text field mentions the replacement patterns being "below" in any case, that's a bit confusing.

**Comment by herbdool** [comment]
@stpaultim @olafgrabienski my fix isn't perfect but it is an improvement. Right now it doesn't show at all for unsaved views. I want able to make it ideal.

**Comment by stpaultim** [comment]
@herbdool & @quicksketch - I'm not sure what to do here. It doesn't feel right to mark this "Works for me," given the inconsistent behavior I experienced. But I do agree with @herbdool that this PR is an improvement. How about this for a status :-)

Tested and this works BETTER for me. (I'll let others manage the tags on this issue)

**Comment by herbdool** [comment]
thanks @stpaultim. Perhaps if we mark this issue as tested and ready then we'll get more eyes and ideas for this. It can always be kicked back down for an even better fix.

**Comment by BWPanda** [comment]
Based on the feedback from others (copied below), the apparent inconsistency of this feature and my own thoughts about the PR, I'm going to recommend that we find a different solution here...

> I found it surprising because the behavior differs from "Rewrite output" where the patterns fieldset is visible all the time. Apart of that, the help in the Custom Text field mentions the replacement patterns being "below" in any case, that's a bit confusing.

> This is an improvement, but it's still inconsistent and a little flakey.

My own thoughts of the PR are that this solution seems inconsistent with other places we do this in Backdrop. Other times, something appears when you tick a box or select an option, or it's just always there. No where else do I recall something appearing only when you start typing in a field (apart from machine names, but they're different). In this case, I think people might want to see the replacement option *before* they start typing...

I appreciate @herbdool's work on this, and his desire to see this improved sooner than later, but I'm going to set this back to 'needs work' in the hope we can fix this for all situations where it happens and remove the inconsistent behaviour.

**Comment by olafgrabienski** [comment]
> Also I've tweaked the PR. Now, if the Global: Custom Text is added it'll show the Replacement Patterns all the time.

That's great news! I've tested it in the PR sandbox website, works for me!

**Comment by quicksketch** [comment]
The current PR has an issue that it causes the token help to show up _always_, even if you're not rewriting a field (@herbdool noted this in a comment: https://github.com/backdrop/backdrop/pull/3014#discussion_r731055464). So I think this still needs some work.

In order to reproduce, I had to follow @stpaultim's directions exactly in https://github.com/backdrop/backdrop-issues/issues/1928#issuecomment-569450466.

With the problem confirmed, I also found that other things did not cause the problem, like if the field you add is the "Global: Custom text" without adding a Body field first.

Turns out what's happening here is that the `#states` value is stored in `Backdrop.settings.states`, a global variable. If you add a field that first sets `#states` on the `$form['alter']['help']` element, then subsequent modal dialogs end up using the same `#states` from the previous field that was added. Changes to `Backdrop.settings` are only additive (not removed), setting new values will overwrite the previous ones but you can't remove the ones that are there already.

Because the "Rewrite the output of this field" checkbox is completely removed on Custom Text fields, the `#states` always evaluates false for global text fields (if any other type of field was added first).

A comprehensive solution would be to somehow remove previous data in `Backdrop.states` for each modal, but I think that could have side-effects. An easier solution is to just change the ID on the elements for the Custom Text form, then the previous `#states` rules won't accidentally match on this element.

I filed a new PR at https://github.com/backdrop/backdrop/pull/3828 that takes this approach.

**Comment by olafgrabienski** [comment]
Great, the PR works for me following these steps:

1. Create a new view of Content, Type = All, Display format = unformatted list of fields
2. Add body field to views. (Default settings)
3. Add "Global: Custom Text" field

Result: the Replacement Patterns fieldset **is** visible (and collapsed), as expected.

Screenshot from the PR sandbox:

![backdrop-pr-3828-view-custom-text](https://user-images.githubusercontent.com/14188307/142404893-d593095d-8b53-4962-98ad-121fcc351a1f.png)

**Comment by herbdool** [comment]
@quicksketch  code looks good. Thanks for digging into this and figuring out the real reason.

**Comment by quicksketch** [comment]
Thank @olafgrabienski and @herbdool for reviewing! I've merged https://github.com/backdrop/backdrop/pull/3828 into 1.x and 1.20.x and closed out @herbdool's https://github.com/backdrop/backdrop/pull/3014.
